source/custom/visa_service.custom.cpp

```diff
@@ -341,21 +341,22 @@ ::grpc::Status VisaService::Read(::grpc::ServerContext* context, const ReadReque
     auto vi_grpc_session = request->vi();
     vi = session_repository_->access_session(vi_grpc_session.name());
     ViUInt32 count = request->count();
-    std::vector<ViByte> buffer(count);
+    std::string* buffer = response->mutable_buffer();
+    buffer->resize(count);
     ViUInt32 return_count{};
-    auto status = library_->Read(vi, buffer.data(), count, &return_count);
+    auto status = library_->Read(vi, (ViByte*)buffer->data(), count, &return_count);
```

Comment:
When in C++ we should use C++ style casts.

Reply:
This is a workaround, reinterpret_cast failed to compile on nilrt.

Reply:
Have you tried `const_cast`? The non-const overload of [basic_string::data](https://en.cppreference.com/w/cpp/string/basic_string/data) was added in C++17.

Reply:
Something like `reinterpret_cast<ViByte*>(const_cast<char*>(buffer->data())`? Is there a easy way to run build on nilrt to try things instead of adding a commit to the PR?

Reply:
I am not a fan of just making C++ happy when it's not readable. The C style cast here is readable.

This is fundamentally a cast from "char*" to "unsigned char*". Yes the problem is that the non-const overload isn't available there in RT. I think casting it twice only makes it confusing. My 2¢ worth.

Reply:
Two other possibilities that should work equally well and would be less confusing to me than above. My preference would be for the first:
`auto status = library_->Read(vi, reinterpret_cast<ViByte*>(&buffer[0]), count, &return_count);`
-or-
`auto status = library_->Read(vi, reinterpret_cast<ViByte*>(&buffer.front()), count, &return_count);`

Reply:
Commit 3

Reply:
On further thought, casting away const here would be a bug: refcounted implementations of `basic_string` don't unshare the string when you call a const method. Using `&buffer[0]` or `&buffer.front()` is correct because these methods have non-const overloads which unshare the string.

Reply:
BTW, calling `&buffer[0]` or `&buffer.front()` on an empty `vector` is undefined behavior, but with `basic_string` it returns a reference to a NUL terminator. C++11 even allows overwriting the NUL terminator (with another NUL terminator): https://en.cppreference.com/w/cpp/string/basic_string/operator_at

Reply:
"Interesting" note about the empty container (shudder). Not an issue here since we first reserve the size.